File: src/services/ticket.ts
import type { Guild, GuildChannel, TextChannel, User } from 'discord.js';
import { ActionRowBuilder, ButtonBuilder, ButtonStyle, ChannelType, PermissionFlagsBits } from 'discord.js';
import { COMPONENT_IDS } from '../utils/constants.js';
import { TicketStatus, TicketType } from '@prisma/client';
import { guildQueries, ticketQueries } from '../database/queries.js';
import { sendLog } from './logging.js';
import { ticketEmbed } from '../utils/embeds.js';
import type { TicketTypeId } from '../utils/constants.js';

const TICKET_CHANNEL_OVERWRITES = (userId: string, supportRoleIds: string[]) => {
  const allowView = [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.ReadMessageHistory, PermissionFlagsBits.AttachFiles, PermissionFlagsBits.EmbedLinks];
  const denyAll = [PermissionFlagsBits.ViewChannel];
  return [
    { id: userId, allow: allowView, deny: [] },
    ...supportRoleIds.map((roleId) => ({ id: roleId, allow: allowView, deny: [] })),
    { id: 'everyone', allow: [], deny: denyAll },
  ];
};

export async function createTicketChannel(
  guild: Guild,
  user: User,
  type: TicketTypeId,
  subject: string | null,
  description: string | null
): Promise<{ channel: TextChannel; ticketNumber: number }> {
  const config = await guildQueries.getOrCreate(guild.id);
  const categoryId = config.ticketCategoryId ?? process.env.TICKET_CATEGORY_ID ?? null;
  const supportRoleIds = guildQueries.getSupportRoleIds(config);
  const ticketNumber = await guildQueries.incrementTicketCounter(guild.id);

  const channelName = `ticket-${String(ticketNumber).padStart(4, '0')}`;
  const channel = await guild.channels.create({
    name: channelName,
    type: ChannelType.GuildText,
    parent: categoryId,
    permissionOverwrites: TICKET_CHANNEL_OVERWRITES(user.id, supportRoleIds),
    topic: `Ticket #${ticketNumber} · ${type} · ${user.tag}`,
  }) as TextChannel;

  await ticketQueries.create({
    guildId: guild.id,
    channelId: channel.id,
    userId: user.id,
    type: type as TicketType,
    subject,
    description,
  });

  const member = await guild.members.fetch(user.id).catch(() => null);
  const displayName = member?.displayName ?? user.username;

  const embed = ticketEmbed({
    ticketType: type,
    subject,
    description,
    userName: displayName,
    userTag: user.tag,
    ticketNumber,
  });

  const row = new ActionRowBuilder<ButtonBuilder>().addComponents(
    new ButtonBuilder().setCustomId(COMPONENT_IDS.TICKET_CLOSE).setLabel('Close ticket').setStyle(ButtonStyle.Danger),
    new ButtonBuilder().setCustomId(COMPONENT_IDS.TICKET_ESCALATE).setLabel('Escalate').setStyle(ButtonStyle.Primary)
  );

  await channel.send({
    content: `${user}, your ticket has been created. Staff will assist you shortly.`,
    embeds: [embed],
    components: [row],
  });

  await sendLog(guild, {
    type: 'ticket_created',
    guild,
    channelId: channel.id,
    user,
    ticketType: type,
    ticketNumber,
  });

  return { channel, ticketNumber };
}

export async function closeTicketChannel(guild: Guild, channel: GuildChannel, closedBy: User): Promise<void> {
  const ticket = await ticketQueries.findByChannel(channel.id);
  if (ticket) {
    const config = await guildQueries.getOrCreate(guild.id);
    await ticketQueries.update(ticket.id, {
      status: TicketStatus.Closed,
      closedAt: new Date(),
      closedBy: closedBy.id,
    });
    await sendLog(guild, {
      type: 'ticket_closed',
      guild,
      channelId: channel.id,
      closedBy,
      ticketNumber: config.ticketCounter,
    });
  }
  await channel.delete('Ticket closed').catch(() => {});
}

export async function reopenTicket(ticketId: string, _reopenedBy: User): Promise<void> {
  await ticketQueries.update(ticketId, {
    status: TicketStatus.Reopened,
    closedAt: null,
    closedBy: null,
  });
}

export async function escalateTicket(ticketId: string, _escalatedBy: User): Promise<void> {
  await ticketQueries.update(ticketId, {
    status: TicketStatus.Escalated,
    escalatedAt: new Date(),
  });
}

File: bot/services/ticket.py
"""Ticket channel create/close/escalate."""
from datetime import datetime, timezone

import discord

from bot.config import TICKET_CATEGORY_ID
from bot.database import queries as db
from bot.services.logging import send_log
from bot.utils.embeds import ticket_embed
from bot.views import TicketChannelView


async def create_ticket_channel(
    guild: discord.Guild,
    user: discord.User,
    ticket_type: str,
    subject: str | None,
    description: str | None,
) -> tuple[discord.TextChannel, int]:
    config = await db.guild_get_or_create(str(guild.id))
    category_id = config.get("ticket_category_id") or TICKET_CATEGORY_ID
    try:
        category_id = int(category_id) if category_id else None
    except (TypeError, ValueError):
        category_id = None
    role_ids = db.get_support_role_ids(config)
    ticket_number = await db.guild_increment_ticket_counter(str(guild.id))

    overwrites = {
        guild.default_role: discord.PermissionOverwrite(view_channel=False),
        user: discord.PermissionOverwrite(
            view_channel=True,
            send_messages=True,
            read_message_history=True,
            attach_files=True,
            embed_links=True,
        ),
    }
    for rid in role_ids:
        try:
            role = guild.get_role(int(rid))
            if role:
                overwrites[role] = discord.PermissionOverwrite(
                    view_channel=True,
                    send_messages=True,
                    read_message_history=True,
                    attach_files=True,
                    embed_links=True,
                )
        except (ValueError, TypeError):
            continue

    channel = await guild.create_text_channel(
        name=f"ticket-{str(ticket_number).zfill(4)}",
        category=discord.Object(id=category_id) if category_id else None,
        overwrites=overwrites,
        topic=f"Ticket #{ticket_number} · {ticket_type} · {user}",
    )

    await db.ticket_create(
        str(guild.id),
        str(channel.id),
        str(user.id),
        ticket_type,
        subject=subject,
        description=description,
    )

    member = guild.get_member(user.id)
    display_name = member.display_name if member else user.name

    embed = ticket_embed(
        ticket_type,
        subject,
        description,
        display_name,
        str(user),
        ticket_number,
    )
    await channel.send(
        content=f"{user.mention}, your ticket has been created. Staff will assist you shortly.",
        embed=embed,
        view=TicketChannelView(),
    )

    await send_log(
        guild,
        "ticket_created",
        channel_id=str(channel.id),
        user=user,
        ticket_type=ticket_type,
        ticket_number=ticket_number,
    )
    return channel, ticket_number


async def close_ticket_channel(
    guild: discord.Guild,
    channel: discord.abc.GuildChannel,
    closed_by: discord.User,
) -> None:
    ticket = await db.ticket_find_by_channel(str(channel.id))
    if ticket:
        await db.ticket_update(
            ticket["id"],
            status="Closed",
            closed_at=datetime.now(timezone.utc),
            closed_by=str(closed_by.id),
        )
        config = await db.guild_get_or_create(str(guild.id))
        await send_log(
            guild,
            "ticket_closed",
            channel_id=str(channel.id),
            closed_by=closed_by,
            ticket_number=config.get("ticket_counter"),
        )
    await channel.delete(reason="Ticket closed")
